### PR TITLE
Add Beat input plugin

### PR DESCRIPTION
**Pull request by nferch**
### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

As discussed in https://github.com/influxdata/telegraf/pull/4911, combine @dmitryilyin 's filebeat and kafkabeat plugins.

### CODE REVIEW

**Comment by nferch** [comment]
@danielnelson any thoughts or feedback?

**Comment by danielnelson** [comment]
I'll have to get back to you on a full review.  Quick look though overall the code is simple and straightforward, though I'd rather avoid FlattenJSON and just pick the values out manually.  I'm also unsure of the utility of collecting the system metrics via this method.  Since all the stats are in one response I also suggest we remove the options about what sections to collect.

**Comment by nferch** [comment]
Filebeat exposes two different URIs, `/` and `/stats`. `/stats` is the only one I really care about, inclusion of the other metrics was based on @esity's comment in https://github.com/influxdata/telegraf/pull/4911 to make a generalized Beats plugin.

Can't speak to much to the implementation as this PR is largely @dmitryilyin's code, but can change from FlattenJSON at the cost of tying the implementation to the API response du jour.

**Comment by nferch** [comment]
I have been using this PR in limited production and it satisfies my use case, detecting queues/backpressure in Filebeat using the `beat_filebeat.events_active` measurement. @dmitryilyin what was your use case? Do you have a Kafka instance to test against?

**Comment by axelrtgs** [comment]
I have been eagerly watching this and waiting for the feature to come to telegraf. I wanted to leave a comment because i believe the current implementation would also work for my use case

**Comment by Esity** [comment]
@danielnelson Just wondering where we are at with this one. Looks pretty good to me

**Comment by nferch** [comment]
@danielnelson Happy New Year! :) Any further thoughts or feedback?

**Comment by axelrtgs** [comment]
@danielnelson Hope all is well, Been eagerly waiting for this to be implemented. Was wondering when you would have a minute to look at it or if there was anything those of us waiting could do to help speed the process along?

**Comment by axelrtgs** [comment]
@danielnelson @sjwang90 I know you guys have a lot on your plate but this seems to be ready since November. Every few months it keeps getting bumped to the next version  Is there anything I can do to get this looked at and make it easier for you to approve and maybe actually get it into the 1.15 release?

**Comment by alsargent** [comment]
Glad we have this on our radar. Just looking at Elastic Beats plugins (aka "community beats"), there are a number of technologies that Telegraf doesn't yet cover: https://www.elastic.co/guide/en/beats/libbeat/current/community-beats.html

**Comment by nferch** [comment]
@sjwang90 re https://github.com/influxdata/telegraf/pull/4911#issuecomment-730862173, any update or end in sight on this PR? It's been open for over a year with several missed milestone inclusions.

Happy to take a look at the code and ensure it's up to date with the current master branch if there's some indication that'd be time worth spending.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:20]: Would it make sense to collect from multiple hosts? Then this should be a list...

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:32]: Please choose a good default e.g. all is collected and keep the number of uncommented sample config options low. The idea is that one can start by just passing the host and might later customize with the options.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:44]: Same. Please comment this option with the (good) default to not confuse new users.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:109]: Make this an `Init()` function and later (in `init()`) pass an empty config.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:131]: How about e.g. proxy settings?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:144]: This default should be set in `Init()`.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:151]: Does it make sense to have a password with an empty username? Does this work at all?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:154]: The local variable `value` might shadow the function argument `value`. To be honest I'm also not sure that I like the passing the result `value` by reference instead of constructing it here and return it. Is there any reason for doing this?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:11]: This is 
```suggestion
	"github.com/influxdata/telegraf/plugins/common/tls"
```
now I think.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:173]: You might directly return `err` here.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:183]: Couldn't you move this out of this function so that we have the same structure below?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:206]: Maybe call `beat.gatherJsonData(infoUrl.String(),...)` here and let `beat.gatherInfoTags(...)` only scrape the tags from the return. This then has all the "shoot a query and parse the return" functions on the same level.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:219]: According to the comment, `FlattenJSON()` does *NOT* handle booleans and strings. Do you maybe want `FullFlattenJSON()` here?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:227]: Maybe you can have one instance of the `flattenerBeat` and clear the fields between the calls...

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:232]: Your metric name `beat_filebeat` does not follow the convention (default: plugin name). This is especially bad if people use `name_override` to customize the output and suddenly don't get the three distinct streams but an undistinguishible mixture.
How about adding the "information type" (`filebeat` in this case) as a tag (e.g. `source="filebeat"`)?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:251]: Same comments as for the `beat.CollectBeatStats` section.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:264]: Could be done in `Init()`.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:276]:
```suggestion
		&Beat{}
```
and setting the defaults in an `Init()` function.

File: plugins/inputs/beat/beat_test.go
Comment (srebhan) [inline plugins/inputs/beat/beat_test.go:19]: Please move this to the individual tests.

File: plugins/inputs/beat/beat_test.go
Comment (srebhan) [inline plugins/inputs/beat/beat_test.go:15]: Please move this to the individual tests.

File: plugins/inputs/beat/beat_test.go
Comment (srebhan) [inline plugins/inputs/beat/beat_test.go:56]: Why is this necessary? The beat-plugin should connect on its own no?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:159]: Can also be moved to `Init()`.

File: plugins/inputs/beat/beat_test.go
Comment (srebhan) [inline plugins/inputs/beat/beat_test.go:34]: The test-data should move to a JSON file you load and dump over HTTP. In the current schema you rely on the compiler to detect that the consts can be compiled out which is... not so nice. :-)

Look at other current test schemes (e.g. into the starlark processor or the `inputs/rabbitmq` tests) to get an idea how this could look like.

File: plugins/inputs/beat/sample_version6.go
Comment (srebhan) [inline plugins/inputs/beat/sample_version6.go:3]: Should be a JSON file. Really. :-)

File: plugins/inputs/beat/sample_version6.go
Comment (srebhan) [inline plugins/inputs/beat/sample_version6.go:13]: Same here.

**Review by srebhan — CHANGES_REQUESTED** [review]
Sorry for this taking soooo long!

I have some comments and like you to consider introducing an `Init()` function to set the defaults and one-time OPs there. This could simplify the code.
Furthermore, please rework the tests. Check e.g. `rabbitmq` on how to keep and use the testdata as JSON files.
Beside this there are some more comments in the code...

Let me know if you have questions.

**Comment by srebhan** [comment]
@nferch any news on this?

**Comment by nferch** [comment]
@srebhan from my cursory review the comments make sense, should have some time this weekend to work on it.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:20]: Think this would be of limited use, I imagine the most common deployments would be a local Telegraf agent and Beat agent per host. So I'd rather address this in a future PR if there's a need that isn't satisfied by a separate configuration per host.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:32]: Will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:44]: Will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:109]: Will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:131]: Couldn't find many other plugins that support proxies, was there something about this one in particular that you thought necessitated it?

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:144]: will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:151]: will fix

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:20]: Fair enough, please make as resolved.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:154]: not sure if I understand the latter part of your comment, this bit is just applying the headers configuration. will rename the variables from the range to avoid them clashing with the value function parameter, LMK if there's other concerns.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:11]: will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:173]: will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:183]: will fix

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:131]: The http plugin supports proxies for examples. The reason is that I had quite some software _not_ accepting proxies in the past and that might be a nightmare in enterprise environments. I don't think adding those is super hard either.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:206]: completely flattened `gatherInfoTags` since it doesn't actually do that much. think this reads much clearer, but open to other ideas.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:219]: will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:227]: only 4 instances so I'll just leave it as-is, but open to changing it if there's considerable efficiency to be gained...

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:154]: What I mean is instead of passing `value` to `gatherJsonData()` you could do `value := &BeatInfo{}` *in* that function and then return `value` to the caller.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:232]: my thinking here was that it made more sense to have separate measurements as there's not much overlap in the responses, it's the interface that is consistent between the endpoints.

don't really have a strong opinion on it though.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:264]: will fix

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:276]: will fix

File: plugins/inputs/beat/beat_test.go
Comment (nferch) [reply plugins/inputs/beat/beat_test.go:19]: will fix

File: plugins/inputs/beat/beat_test.go
Comment (nferch) [reply plugins/inputs/beat/beat_test.go:15]: will fix

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:227]: Hmm. I don't see the reason not to declare `flattener := jsonparser.JSONFlattener{}` and use it in the `if`s. The flattener is even stateless, so you save the construction lines... :-)

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:159]: I don't understand, this is setting `Host` on the HTTP request object, which is instantiated per `gatherJsonData` call.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:232]: Go ahead as-is for now. I have to discuss this more with Influx-people... Many plugins do this and if you don't be careful you easily screw up the metrics when using `name_override` as it doesn't care about the suffix...
Mark resolved please.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:159]: You are absolutely right. Sorry missed this one. Mark as resolved.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:131]: From what I can tell, the http plugin is the only other plugin that supports proxies. I could see adding it if there's enough prevailing need or consistency with the other plugins that support it. Seems like something that could be easily added if necessary.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:154]: ahh yeah, that is a bit weird. I'm not the original author so I don't have the full context as far as history or intentions, but I'll fix this one.

Comment (nferch) [reply plugins/inputs/beat/beat.go:154]: actually, I understand now. :) It's passed a `BeatInfo` and later a `BeatStats`, so it needs to accept an `interface{}`

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:227]: I'm not too familiar with the API, but that doesn't seem to be the case. Simply making the additional calls use the same instance without clearing it (is there a convenient way to do that?) causes the tests to fail.

File: plugins/inputs/beat/beat_test.go
Comment (nferch) [reply plugins/inputs/beat/beat_test.go:56]: will fix

File: plugins/inputs/beat/beat_test.go
Comment (nferch) [reply plugins/inputs/beat/beat_test.go:34]: will fix

File: plugins/inputs/beat/sample_version6.go
Comment (nferch) [reply plugins/inputs/beat/sample_version6.go:3]: will fix

File: plugins/inputs/beat/sample_version6.go
Comment (nferch) [reply plugins/inputs/beat/sample_version6.go:13]: will fix

**Comment by nferch** [comment]
@srebhan ok, I think I addressed all of your comments. LMK if there's anything I missed.

Would welcome any other patches or real-world testing, as I'm not the original author of the code and only have Filebeat running to test it against.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:84]: Why do you need those to be pointers?

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:106]: Please don't declare them here. Either you declare those in `Init()` or you directly set the defaults there.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:84]: Since we are setting defaults in `Init()` after the config has been read, it would not be possible to discern a false setting from an unset setting. Open to other ways to achieve this, this is how I've seen it done in other projects.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:106]: Was necessary to make them vars in order to take the address of them. Thought it was cleaner to separate the default values from the code of setting the defaults.

File: plugins/inputs/beat/README.md
Comment (srebhan) [inline plugins/inputs/beat/README.md:32]: You might also want to comment the default to match `sampleConfig`.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:110]:
```suggestion
	beat.client, err := beat.createHTTPClient()
```

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:162]: I take your word on this! ;-) If you really don't want to do it now, mark this as resolved...

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:192]: You can also do
```suggestion
	return json.NewDecoder(response.Body).Decode(value)
```
Both ok for me.

File: plugins/inputs/beat/README.md
Comment (nferch) [reply plugins/inputs/beat/README.md:32]: good catch, will fix.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:207]: Please mark as resolved.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:227]: Still open...

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:110]: unless I'm mistaken, it's necessary to declare separately since `:=` will only work for new variables or variables defined in the current block. The suggested change does not compile.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [inline plugins/inputs/beat/beat.go:32]: Concerning our "default value" discussion below, it might be easier to allow the user to specify a list, i.e.
```
collect =["beat", "libbeat", "filebeat"]
```
This way you know if an option was selected (it's in the list) or not. Maybe you can define an empty list to "collect all"...

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:110]: Hmmm what is the error message? It should work I think...

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:32]: This is also more extensible in the future if you want to e.g. increase granularity...

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:106]: See my comment to change to a collection list in the form `collect = ["beat", "libbeat",...]`. You can use `internal/choice` to check the options an see what is in the user-setting.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:110]: `plugins/inputs/beat/beat.go:109:6: non-name beat.client on left side of :=`

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:32]: I don't have a strong opinion on this, but don't feel much motivation to change it either. The design decision to use booleans predated my commits which sought to integrate the plugins in the simplest way possible. The use of booleans to toggle functionality seemed similar to other plugins so it didn't seem unusual to use them in this way. I don't see how it would make a difference to the end user either, but I obviously have a limited perspective on this.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:110]: Sorry, missed the structure part... Should sleep more. :-( Keep as is and mark my this one as resolved.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:32]: Yeah it's similar to the plugins where we currently have a lot of trouble extending them without breaking configs... Many plugins therefore move to `include`, `exclude` lists to allow sophisticated filtering of metrics. This is nice for the end-user for cases where you want all but one.

Comment (nferch) [reply plugins/inputs/beat/beat.go:32]: So is this just a "nice to have" or a blocker for moving forward? I can see how it'd be useful for some more complicated use cases or extensibility, but I don't have much reason to believe that how it is currently implemented doesn't solve the problems of the present. And I'm reluctant to spend much time solving the problems of the future, when they can be easily addressed via a PR once they materialize.

I'd assume that the prospective users of this plugin would prefer having a potentially less-perfect solution sooner than a more future proof solution later, but I defer to anyone else who is following this thread to chime in with their use cases and preferences, as all I have to go by on is my personal one.

Comment (axelrtgs) [reply plugins/inputs/beat/beat.go:32]: I would vote for nice to have vs blocker. i know i would love to see this out so i can monitor filebeats status even if its not perfect. Although if its not too much work it may be reasonable to do it now since i know fixing things later rarely happens and although i would love to contribute i have 0 knowledge of GO to even pick this up myself in a second PR.

Perhaps merging this functional PR as is with the non perfect config but then opening an issue or second PR to enhance the config may be a compromise between both options.

File: plugins/inputs/beat/beat.go
Comment (ssoroka) [reply plugins/inputs/beat/beat.go:84]: set the defaults in init(), and then you don't have to check them at all, and the fields don't have to be pointers.

Comment (nferch) [reply plugins/inputs/beat/beat.go:84]: @ssoroka This is how it was prior to the comments on https://github.com/influxdata/telegraf/pull/6653#discussion_r530877623

Comment (ssoroka) [reply plugins/inputs/beat/beat.go:84]: Sorry about that. the correct way to set defaults is when you create the plugin, (around the lowercase `init()` call), By the time `Init()` is called, the config has been loaded over top of the defaults, and has the correct settings. Trying to set them in `Init()` leads to some weirdness, as you've experienced. I'd also like to go ahead with the change to change the bools to a `[]string` slice, because once we merge it the config is locked and we can't break backwards compatibility.

**Comment by srebhan** [comment]
Here is my [implementation](https://github.com/influxdata/telegraf/files/5705290/beat-plugin-include-list.patch.txt)  of converting the booleans to a list of includes... Feel free to use it.

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:32]: The problem is that we **cannot** fix this *later* as we need to keep backward compatible with the config. Instead of arguing over this again I implemented what is necessary in less time than the discussion took so far...
Please see my comment below with the attached patch (`git apply ...`). Feel free to use it!

File: plugins/inputs/beat/beat.go
Comment (srebhan) [reply plugins/inputs/beat/beat.go:84]: @ssoroka this is my fault. I thought it is cleaner to initialize the defaults in `Init()` but completely missed that you cannot know if the boolean was specified as false by the user or it is false due to not being initialized. Again, my fault. Sorry @nferch for the confusion I caused!

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:32]: Thanks, applied.

File: plugins/inputs/beat/beat.go
Comment (nferch) [reply plugins/inputs/beat/beat.go:84]: Ok, I've reverted the flow to set defaults in `init()`.

**Comment by nferch** [comment]
@srebhan @ssoroka I believe I've addressed all the comments and feedback. LMK if there's anything else, otherwise I'll squash and rebase so it's ready for merge.

**Review by srebhan — APPROVED** [review]
Looks good to me now! Also thx for further cleanup!

**Comment by srebhan** [comment]
@ssoroka can you maybe add a second opinion as I provided the include-list-part before flagging this ready-to-merge!?

**Review by ssoroka — APPROVED** [review]
lgtm :)

**Comment by ssoroka** [comment]
thanks for the epic effort here. much appreciated

**Pull request merged**
